src/bitset.rs

```diff
@@ -295,6 +312,161 @@ impl<T> Iterator for Iter<T>
+            self.layer2.extend(repeat(0).map(|_| AtomicUsize::new(0)).take(count));
+        }
+        if self.layer1.len() <= p1 {
+            let count = p1 - self.layer1.len() + 1;
+            self.layer1.extend(repeat(0).map(|_| AtomicUsize::new(0)).take(count));
+        }
+        if self.layer0.len() <= p0 {
+            let count = p0 - self.layer0.len() + 1;
+            self.layer0.extend(repeat(0).map(|_| AtomicUsize::new(0)).take(count));
+        }
+    }
+
+    /// Adds `id` to the `AtomicBitSet`. Returns `true` if the value was
+    /// already in the set.
+    ///
+    /// Because we cannot safely extend an AtomicBitSet without unique ownership
+    /// this will panic if the Index is out of range
+    #[inline]
+    pub fn add_atomic(&self, id: Index) -> bool {
+        let (p0, p1, p2) = offsets(id);
+        let old = self.layer0[p0].fetch_or(id.mask(SHIFT0), Ordering::Relaxed);
+        if old & id.mask(SHIFT0) != 0 {
+            return true;
+        }
+
+        self.layer1[p1].fetch_or(id.mask(SHIFT1), Ordering::Relaxed);
+        self.layer2[p2].fetch_or(id.mask(SHIFT2), Ordering::Relaxed);
+        self.layer3.fetch_or(id.mask(SHIFT3), Ordering::Relaxed);
+        false
+    }
+
+    /// Adds `id` to the `BitSet`. Returns `true` if the value was
+    /// already in the set.
+    #[inline]
+    pub fn add(&mut self, id: Index) -> bool {
+        if id.offset(SHIFT1) >= self.layer0.len() {
+            self.extend(id);
+        }
+
+        self.add_atomic(id)
+    }
+
+    /// Removes `id` from the set, returns `true` if the value
+    /// was removed, and `false` if the value was not set
+    /// to begin with.
+    #[inline]
+    pub fn remove(&mut self, id: Index) -> bool {
+        let (p0, p1, p2) = offsets(id);
+
+        if p0 >= self.layer0.len() {
+            return false;
+        }
+
+        if self.layer0[p0].load(Ordering::Relaxed) & id.mask(SHIFT0) == 0 {
+            return false;
+        }
+
+        // if the bitmask was set we need to clear
+        // its bit from layer0 to 3. the layers abover only
+        // should be cleared if the bit cleared was the last bit
+        // in its set
+        self.layer0[p0].fetch_and(!id.mask(SHIFT0), Ordering::Relaxed);
+        if self.layer0[p0].load(Ordering::Relaxed) != 0 {
+            return true;
+        }
+
+        self.layer1[p1].fetch_and(!id.mask(SHIFT1), Ordering::Relaxed);
+        if self.layer1[p1].load(Ordering::Relaxed) != 0 {
+            return true;
+        }
+
+        self.layer2[p2].fetch_and(!id.mask(SHIFT2), Ordering::Relaxed);
+        if self.layer2[p2].load(Ordering::Relaxed) != 0 {
+            return true;
+        }
+
+        self.layer3.fetch_and(!id.mask(SHIFT3), Ordering::Relaxed);
+        return true;
+    }
+
+    /// Returns `true` if `id` is in the set.
+    #[inline]
+    pub fn contains(&self, id: u32) -> bool {
+        let p0 = id.offset(SHIFT1);
+        if p0 >= self.layer0.len() {
```

Comment:
could merge with the following condition for clarity, as in `p0 < self.layer0.len() && ...`